internal/class.h

```diff
@@ -79,103 +135,355 @@ STATIC_ASSERT(shape_max_variations, SHAPE_MAX_VARIATIONS < (1 << (sizeof(((rb_cl
-#define RCLASS_SUPERCLASSES_INCLUDE_SELF FL_USER2
-#define RICLASS_ORIGIN_SHARED_MTBL FL_USER3
+// 1 is for RUBY_FL_SINGLETON or RMODULE_ALLOCATED_BUT_NOT_INITIALIZED (see class.c)
+#define RCLASS_PRIME_CLASSEXT_WRITABLE FL_USER2
+// 3 is RMODULE_IS_REFINEMENT for RMODULE
+// 4-19: SHAPE_FLAG_MASK
 
-static inline st_table *
-RCLASS_FIELDS_HASH(VALUE obj)
+/* class.c */
+rb_classext_t * rb_class_duplicate_classext(rb_classext_t *orig, VALUE obj, const rb_namespace_t *ns);
+void rb_class_ensure_writable(VALUE obj);
+
+static inline int
+RCLASS_SET_NAMESPACE_CLASSEXT(VALUE obj, const rb_namespace_t *ns, rb_classext_t *ext)
 {
-    RUBY_ASSERT(RB_TYPE_P(obj, RUBY_T_CLASS) || RB_TYPE_P(obj, RUBY_T_MODULE));
-    RUBY_ASSERT(rb_shape_obj_too_complex_p(obj));
-    return (st_table *)RCLASS_FIELDS(obj);
+    int first_set = 0;
+    st_table *tbl = RCLASS_CLASSEXT_TBL(obj);
+    VM_ASSERT(NAMESPACE_USER_P(ns)); // non-prime classext is only for user namespace, with ns_object
+    VM_ASSERT(ns->ns_object);
+    VM_ASSERT(RCLASSEXT_NS(ext) == ns);
+    if (!tbl) {
+        RCLASS_CLASSEXT_TBL(obj) = tbl = st_init_numtable_with_size(1);
+    }
+    if (rb_st_table_size(tbl) == 0) {
+        first_set = 1;
+    }
+    rb_st_insert(tbl, (st_data_t)ns->ns_object, (st_data_t)ext);
+    return first_set;
+}
+
+static inline bool
+RCLASS_PRIME_CLASSEXT_READABLE_P(VALUE klass)
+{
+    VM_ASSERT(RB_TYPE_P(klass, T_CLASS) || RB_TYPE_P(klass, T_MODULE) || RB_TYPE_P(klass, T_ICLASS));
+    // if the lookup table exists, then it means the prime classext is NOT directly readable.
+    return RCLASS_CLASSEXT_TBL(klass) == NULL;
+}
+
+static inline bool
+RCLASS_PRIME_CLASSEXT_WRITABLE_P(VALUE klass)
+{
+    VM_ASSERT(RB_TYPE_P(klass, T_CLASS) || RB_TYPE_P(klass, T_MODULE) || RB_TYPE_P(klass, T_ICLASS));
+    return FL_TEST(klass, RCLASS_PRIME_CLASSEXT_WRITABLE);
 }
 
 static inline void
-RCLASS_SET_FIELDS_HASH(VALUE obj, const st_table *tbl)
+RCLASS_SET_PRIME_CLASSEXT_WRITABLE(VALUE klass, bool writable)
 {
-    RUBY_ASSERT(RB_TYPE_P(obj, RUBY_T_CLASS) || RB_TYPE_P(obj, RUBY_T_MODULE));
-    RUBY_ASSERT(rb_shape_obj_too_complex_p(obj));
-    RCLASS_FIELDS(obj) = (VALUE *)tbl;
+    VM_ASSERT(RB_TYPE_P(klass, T_CLASS) || RB_TYPE_P(klass, T_MODULE) || RB_TYPE_P(klass, T_ICLASS));
+
+    if (writable) {
+        FL_SET(klass, RCLASS_PRIME_CLASSEXT_WRITABLE);
+    }
+    else {
+        FL_UNSET(klass, RCLASS_PRIME_CLASSEXT_WRITABLE);
+    }
 }
 
-static inline uint32_t
-RCLASS_FIELDS_COUNT(VALUE obj)
+static inline rb_classext_t *
+RCLASS_EXT_TABLE_LOOKUP_INTERNAL(VALUE obj, const rb_namespace_t *ns)
 {
-    RUBY_ASSERT(RB_TYPE_P(obj, RUBY_T_CLASS) || RB_TYPE_P(obj, RUBY_T_MODULE));
-    if (rb_shape_obj_too_complex_p(obj)) {
-        uint32_t count;
+    st_data_t classext_ptr;
+    st_table *classext_tbl = RCLASS_CLASSEXT_TBL(obj);
+    if (classext_tbl) {
+        if (rb_st_lookup(classext_tbl, (st_data_t)ns->ns_object, &classext_ptr)) {
```

Comment:
It doesn't seem like this is safe to do without holding the VM lock. (The VM lock is held on write so it must also be held on read, otherwise the lock has no benefit)